ethcore/src/client/client.rs

```diff
@@ -909,29 +909,29 @@ impl Client {
 	}
 
 	fn build_last_hashes(&self, parent_hash: &H256) -> Arc<LastHashes> {
-		{
-			let hashes = self.last_hashes.read();
-			if hashes.front().map_or(false, |h| h == parent_hash) {
-				let mut res = Vec::from(hashes.clone());
-				res.resize(256, H256::zero());
-				return Arc::new(res);
-			}
-		}
-		let mut last_hashes = LastHashes::new();
-		last_hashes.resize(256, H256::zero());
-		last_hashes[0] = parent_hash.clone();
-		let chain = self.chain.read();
-		for i in 0..255 {
-			match chain.block_details(&last_hashes[i]) {
-				Some(details) => {
-					last_hashes[i + 1] = details.parent.clone();
-				},
-				None => break,
+		let mut seen_hashes = self.last_hashes.write();
```

Comment:
Note this makes this method more or less single-threaded, the reasoning is that `seen_hashes.front()` should only be queried if the thread has mutual exclusion. Thus, it might be slower than before.

Reply:
What calls `build_last_hashes`? Like, is it on the hot path?

Reply:
Seems to be on a hot path, for example `check_and_lock_block` calls `build_last_hashes`.

Called every time a block is ready for insertion, so I don't know really

Reply:
@tomusdrw @debris Do you have any thoughts on this? Do you think the impact is severe?